rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -0,0 +1,148 @@
+
+        if (committedCache.containsKey(wrappedKey)) {
+            return committedCache.get(wrappedKey);
+        }
+        if (uncommittedCache.containsKey(wrappedKey)) {
+            return uncommittedCache.get(wrappedKey);
+        }
+
+        byte[] value = base.get(key);
+        //null value, as expected, is allowed here to be stored in committedCache
+        committedCache.put(wrappedKey, value);
+        return value;
+    }
+
+    @Override
+    public byte[] put(byte[] key, byte[] value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        // here I could check for equal data or just move to the uncommittedCache.
+        byte[] priorValue = committedCache.get(wrappedKey);
+        if (priorValue != null && Arrays.equals(priorValue, value)) {
+            return value;
+        }
+
+        committedCache.remove(wrappedKey);
+        uncommittedCache.put(wrappedKey, value);
+        return value;
+    }
+
+    @Override
+    public void delete(byte[] key) {
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+        committedCache.remove(wrappedKey);
+        uncommittedCache.put(wrappedKey, null);
+    }
+
+    @Override
+    public Set<byte[]> keys() {
+        Stream<ByteArrayWrapper> baseKeys = base.keys().stream().map(ByteArrayWrapper::new);
+        Stream<ByteArrayWrapper> committedKeys = committedCache.entrySet().stream()
+                .filter(e -> e.getValue() != null)
+                .map(Map.Entry::getKey);
+        Stream<ByteArrayWrapper> uncommittedKeys = uncommittedCache.entrySet().stream()
+                .filter(e -> e.getValue() != null)
+                .map(Map.Entry::getKey);
+        Set<ByteArrayWrapper> uncommittedKeysToRemove = uncommittedCache.entrySet().stream()
+                .filter(e -> e.getValue() == null)
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toSet());
+        Set<ByteArrayWrapper> knownKeys = Stream.concat(Stream.concat(baseKeys, committedKeys), uncommittedKeys)
+                .collect(Collectors.toSet());
+        knownKeys.removeAll(uncommittedKeysToRemove);
+
+        return knownKeys.stream()
+                .map(ByteArrayWrapper::getData)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public void updateBatch(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> keysToRemove) {
+        if (rows.containsKey(null) || rows.containsValue(null)) {
+            throw new IllegalArgumentException("Cannot update null values");
+        }
+
+        // don't move unchanged keys unnecessarily
+        Map<ByteArrayWrapper, byte[]> rowsToUpdate = new HashMap<>(rows);
+        rowsToUpdate.entrySet().removeIf(e -> Arrays.equals(e.getValue(), committedCache.get(e.getKey())));
+
+        committedCache.keySet().removeAll(rowsToUpdate.keySet());
```

Comment:
Rephrasing the previous comment with pseudocode: here the code that follows should be:
**keysToCommitRemove = keysToRemove  excluding commitCache.keys(filter = value == null).**
committedCache.keySet().removeAll(keysToRemove);
uncommittedCache.putAll(rowsToUpdate);
**keysToCommitRemove**.forEach((keyToRemove) -> 
 uncommittedCache.put(keyToRemove, null));
to get optimum cache strategy.
But all this key filtering would not be needed if updateBatch() received a single list, with null values for deletions, as was in the original.

Reply:
Philosophic side note:  There should be a balance between the simplicity of an interface vs the simplicity of an implementation. If a simple interface leads to a complex implementation, then perhaps it's better to have a no-so-simple interface, with straightforward implementation.

Reply:
Final suggested code:
public void updateBatch(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> keysToRemove) {
         if (rows.containsKey(null) || rows.containsValue(null)) {
             throw new IllegalArgumentException("Cannot update null values");
  Map<ByteArrayWrapper, byte[]> combinedRows= new HashMap<>(rows);
 keysToRemove.forEach((keyToRemove) -> combinedRows.put(keyToRemove, 
 null));
 combinedRows.entrySet().removeIf(e -> Arrays.equals(e.getValue(), 
  uncommittedCache.get(e.getKey())));
 committedCache.removeAll(combinedRows);
 uncommitedCache.putAll(combinedRows);
}